### PR TITLE
Add a post about ndindex

### PR DESCRIPTION
**Pull request by asmeurer**
Please review.

This also adds the commonmark extension, as it wasn't possible to format some of this with the default Markdown. Unfortunately, this disables syntax highlighting in the code cells. We can add highlight.js to the post if we want to re-add it. I played with it a bit but couldn't get it to get colors that looked good, so I didn't bother with it.

### CODE REVIEW

**Comment by asmeurer** [comment]
CC @scopatz

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 1]
```diff
@@ -0,0 +1,325 @@
+<!--
+.. title: Introducing ndindex, a Python library for manipulating indices of ndarrays
+.. slug: introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays
+.. date: 2020-04-09 15:42:24 UTC-05:00
+.. tags:
+.. category:
+.. link:
+.. description:
+.. type: text
```
Comment (scopatz) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:9]: Is there not an author entry?

**Comment by scopatz** [comment]
LGTM.  Thanks!

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown (hunk 1 above)
Comment (asmeurer) [reply posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:6]: Oh yeah. I guess it doesn't ask you for it. I added some metadata. We will also need to update the date whenever this gets merged.

**Comment by rgommers** [comment]
Thanks @asmeurer!

> This also adds the commonmark extension, as it wasn't possible to format some of this with the default Markdown.

Can you comment on what didn't work? I didn't spot anything unusual in your post.

> Unfortunately, this disables syntax highlighting in the code cells. We can add highlight.js to the post if we want to re-add it. I played with it a bit but couldn't get it to get colors that looked good, so I didn't bother with it.

I also don't see why highlighting is broken in your post, but still works in older posts. What am I missing?

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+  array([2, 3, 4, 5, 6, 7, 8])
+  ```
+
+The following things are not yet implemented, but are planned.
+
+- `idx.newshape(shape)` returns the shape of `a[idx]`, assuming `a` has shape
+  `shape`.
+
+- `ellipsis`, `Newaxis`, `IntegerArray`, and `BooleanArray` types, so that all
+  types of indexing are support.
```
Comment (rgommers) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:143]: typo, should be "are supported"

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+- `i1[i2]` will create a new ndindex `i3` (when possible) so that
+  `a[i1][i2] == a[i3]`.
+
+- `split(i0, [i1, i2, ...])` will return a list of indices `[j1, j2, ...]`
+  such that `a[i0] = concat(a[i1][j1], a[i2][j2], ...)`
+
+- `i1 + i2` will produce a single index so that `a[i1 + i2]` gives all the
+  elements of `a[i1]` and `a[i2]`.
+
+- Support [NEP 21 advanced
```
Comment (rgommers) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:154]: advanced --> "outer indexing and vectorized indexing"?

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+And more. If there is something you would like to see this library be able to
+do, please [open an issue](https://github.com/quansight/ndindex/issues). Pull
+requests are welcome as well.
+
+## Testing and correctness
+
+The most important priority for a library like this is correctness. Index
+manipulations, and especially slice manipulations, are complicated to code
+correctly, and the code for them typically involves dozens of different
+branches for different cases and formulas that .
```
Comment (rgommers) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:166]: sentence not finished

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+  take billions of years to complete.
+
+- **Hypothesis tests.**
+  [Hypothesis](https://hypothesis.readthedocs.io/en/latest/index.html) is a
+  library that can intelligently check a combinatorial search space of inputs.
+  This requires writing Hypothesis strategies that can generate all the
+  relevant types of indices. All ndindex tests have Hypothesis tests, even if
+  they are also tested exhaustively.
+
+  The Hypotheses test for the above test looks like this
```
Comment (rgommers) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:238]: "Hypotheses" -> "Hypothesis"

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+          # shape == ()
+          return
+      assert reduced.start >= 0
+      # We cannot require stop > 0 because if stop = None and step < 0, the
+      # only equivalent stop that includes 0 is negative.
+      assert reduced.stop != None
+      assert len(reduced) == len(a[reduced.raw]), (s, shape)
+  ```
+
+  In order to tell Hypotheses how to search the example space, we must define
```
Comment (rgommers) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:284]: "Hypotheses" -> "Hypothesis"

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+  some functions to tell it how to draw example objects of a given type, in
+  this case, slices and shape parameters for NumPy arrays. These strategies,
+  as they are called, can be reused for multiple tests. Hypothesis then
+  automatically and intelligently draws examples from the sample space to try
+  to find one that fails the test. You can think of Hypotheses as a fuzzer, or
```
Comment (rgommers) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:289]: once more "Hypotheses" -> "Hypothesis"

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+failure examples. When an exhaustive test fails, the failure will always be
+from the first set of inputs in the loop that produces a failure. Hypothesis
+on the other hand attempts to "shrink" the failure input to smallest input
+that still fails. For example, a failing exhaustive slice test might give
+`Slice(-10, -9, -10)` as a the failing example, but Hypothesis would shrink it
+to `Slice(-2, -1, -1)`.
+
+Another reason for the duplication is that Hypothesis can sometimes test a
+slightly expanded test space without any additional consequences. For example,
+the above Hypotheses tests all types of array shapes, whereas the exhaustive
```
Comment (rgommers) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:304]: and last one "Hypotheses" -> "Hypothesis"

File: posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+test tests only 1-dimensional shapes. This doesn't affect things because
+Hypotheses will always shrink large shapes to a 1-dimensional shape in the
```
Comment (rgommers) [inline posts/2020/04/introducing-ndindex-a-python-library-for-manipulating-indices-of-ndarrays.markdown:306]: oh no, one more here

**Comment by rgommers** [comment]
Nice post, thanks Aaron. It reads well, all my comments are really minor.

**Comment by asmeurer** [comment]
> Can you comment on what didn't work? I didn't spot anything unusual in your post.

The code blocks nested under the bullet points. If you know how to make them work in markdown we can switch it to that.

**Comment by asmeurer** [comment]
> I also don't see why highlighting is broken in your post, but still works in older posts. What am I missing?

Commonmark doesn't go through pygments like python-markdown does. Instead, if you want highlighting, you have to do it yourself, for instance, with highlight.js.

**Comment by rgommers** [comment]
> > I also don't see why highlighting is broken in your post, but still works in older posts. What am I missing?
> 
> Commonmark doesn't go through pygments like python-markdown does. Instead, if you want highlighting, you have to do it yourself, for instance, with highlight.js.

Yeah but everything gets re-rendered. So maybe `python-markdown` is used first, and on failure `commonmark` is used?

**Comment by asmeurer** [comment]
I configured it so that python-markdown is used on .md files and commonmark is used on .markdown files, so this shouldn't break any of the existing posts.

**Comment by asmeurer** [comment]
Sorry I was off last week. I addressed the review comments. This is ready to go from my end.

**Comment by rgommers** [comment]
Merged, live it goes:) Thanks Aaron, this is a really nice post (and cool library)!

**Pull request merged**